Overhaul README

azrogers opened the pull request:
Closes #984. As part of our documentation effort, the Cesium Native README needs to be updated as the project has grown and changed focus. To list some changes in this PR:

- Screenshots added, sourced from existing material on Cesium Native's integrations, to show off Cesium Native and grab attention.
- Libraries overview updated to include libraries that were missing.
- Diagram showing how Cesium Native integrates into the overall Cesium ecosystem has been updated.
- Dependency diagrams are now being generated. This is something CMake can do itself, but the GraphViz output it creates is a mess and either includes too many third party dependencies, or no third party dependencies at all. I've written a script that takes this GraphViz output, prunes it, and turns it into a Mermaid diagram, one for each target and one for the project overall.
- Contributor information has been added, adapted from CesiumJS's `CONTRIBUTORS.md`. We may want to put this in a different file as they've done.
- Table of contents has been included.
- Added a list of projects using Cesium Native, including community projects.
- Added Cesium3DTilesContent to Doxygen - apparently this got missed!
- Made sure the resulting README displays the same in Doxygen as it does on GitHub, as it's now the landing page of the documentation.

kring reviewed (commented):
This looks great @azrogers! The README - and this content generally - will continue to evolve, but this a big step in the right direction. Nicely done with the dependency graph generation! I have a few specific comments below.

j9liu commented on README.md:
```diff
@@ -1,43 +1,135 @@
+  <tbody>
+    <tr>
+      <td>
+
```
Nitpick, but this extra line break is noticably offsetting this table cell :')

![image](https://github.com/user-attachments/assets/21d44174-7765-4a1b-be0d-6fbd13969b53)

j9liu commented on README.md:
````diff
@@ -123,3 +215,78 @@ cd ../..
+```
+* From the `tools/dep-graph-gen` directory, run `npm run generate-dep-graph` to regenerate the graphs.
+
+## 🤝Contributing
````
Could the "Contributing" sections be moved to a `CONTRIBUTING.md` file in the `doc` folder, similar to how it's organized in CesiumJS? We can still link to it similar to how CesiumJS's readme does [here](https://github.com/CesiumGS/cesium?tab=readme-ov-file#what-next)

j9liu commented on README.md:
```diff
@@ -1,43 +1,135 @@
+        <p align="center"><sub>The Copernicus Crater, visualized with Cesium Moon Terrain in Cesium for Unreal.</sub></p>
+      </td>
+      <td>
+        ![San Francisco](doc/img/screenshot-san-francisco.jpg)
```
Is this the Aerometrex San Francisco dataset? I think it is since I'm not seeing any Google attribution, and in that case, we should credit them.

azrogers replied:
I believe so. I grabbed it from https://cesium.com/learn/omniverse/omniverse-quickstart/, where it doesn't have any attribution listed. But looking in the later tutorials, it does seem to be the Aerometrex San Francisco dataset.

j9liu commented on README.md:
```diff
@@ -1,43 +1,135 @@
+      <td>
+
+![Googleplex](doc/img/screenshot-googleplex.jpg)
+        <p align="center"><sub>The Googleplex in Mountain View, California, USA, visualized with Photorealistic 3D Tiles in Cesium for Unity.</sub></p>
```
```suggestion
        <p align="center"><sub>The Googleplex in Mountain View, California, USA, visualized with Google Photorealistic 3D Tiles in Cesium for Unity.</sub></p>
```

j9liu commented on README.md:
```diff
@@ -1,43 +1,135 @@
+
+Cesium Native relies on various third-party packages to accomplish tasks like image decoding, URI parsing, and math operations. These dependencies are automatically obtained and built using [vcpkg](https://vcpkg.io/en/) when building through CMake. Below is a graph of all the Cesium Native libraries and their dependencies on third party packages and each other.
+
+![](doc/img/dependency-graph.svg)
```
Seconding that it's a bit small and hard to read -- is it possible to **bold** the text at all, or increase the line thickness of the nodes? That could help with readability

azrogers replied:
I can definitely do either one of those, or both, or any other styling change - but I think the problem with it just being too large of a graph to embed in the README is going to remain. Ideally, there would be a pan-and-zoom feature like the Mermaid live editor has, but this seems like more trouble than it's worth to set up for Doxygen, and it's just not an option for GitHub.

j9liu commented on README.md:
```diff
@@ -1,43 +1,135 @@
+
+![](doc/img/dependency-graph.svg)
+
 ## 💻Developers
```
I kind of like what we did in Cesium for Unreal's README where we point developers to a more dedicating landing page for development: https://github.com/CesiumGS/cesium-unreal?tab=readme-ov-file#computer-developing-with-unreal-engine

That might be personal preference though -- I like compact READMEs. 😛 Does anyone else have opinions?

azrogers commented on CMakeGraphVizOptions.cmake:
```diff
@@ -0,0 +1,7 @@
+set(GRAPHVIZ_GENERATE_PER_TARGET FALSE)
```
I don't believe so. The file needs to be in `CMAKE_BINARY_DIR` or `CMAKE_SOURCE_DIR` so it can be automatically loaded when CMake is run with the `--graphviz` option,

azrogers commented on CesiumGeospatial/CMakeLists.txt:
```diff
@@ -43,6 +43,7 @@ PUBLIC
     earcut
 PRIVATE
     s2::s2
+    absl::string_view
```
Since `absl::string_view` is used from the `S2CellID` class, I'd assumed it was an omission. However, looking more closely, `absl::string_view` comes as a dependency of `s2` itself, so this is unnecessary.

azrogers commented on doc/CMakeLists.txt:
```diff
@@ -86,6 +88,9 @@ if(DOXYGEN_FOUND)
     set(DOXYGEN_EXCLUDE_PATTERNS "*/node_modules/*")
     set(DOXYGEN_IMAGE_PATH "./")
+    set(DOXYGEN_MARKDOWN_ID_STYLE GITHUB)
+    set(DOXYGEN_DOT_IMAGE_FORMAT svg)
+    set(DOXYGEN_INTERACTIVE_SVG YES)
```
I was hoping they'd do something. Since the dependency graph is an SVG, it would be nice to get the pan-and-zoom features when embedding it. But it seems like the setting only applies to DOT charts and not to other SVGs.

azrogers commented on doc/diagrams/dependencies/Cesium3DTiles.mmd:
```diff
@@ -0,0 +1,10 @@
+---
+config:
+  layout: elk
```
Ah, this setting is really for the full dependency graph first and foremost, and when generating an SVG with Mermaid CLI it worked as-is. Using the ELK layout algorithm helps the graph from becoming an ugly mess. Mermaid's built-in layout algorithm generates a bunch of curvy lines that overlap each other and make the chart difficult to read. ELK makes those nice orderly straight lines, like the wiring on a PCB.

azrogers commented on README.md:
```diff
@@ -1,43 +1,135 @@
+<!-- omit in toc -->
```
The main goal for me actually wasn't Doxygen or GitHub, but the [Markdown All in One VS Code extension](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one#table-of-contents). Helped me generate it and keep it updated while I was writing. However, the point you're making here is a good one. The limitations of trying to make one Markdown file that works on both Doxygen and GitHub is why, for example, there's weird indentation and formatting in the source of the first row of the screenshots table - Doxygen won't resolve images properly unless they're using Markdown image tags, and Markdown image tags are picky about where exactly they're placed.

azrogers commented:
Handled a number of the review comments, but it seems like the big question we have remaining is splitting up the README. Splitting it up into a README, a CONTRIBUTING, and a Developers Guide, and splitting the Doxygen README from the GitHub README. Does this sound good to everyone? I'm personally indifferent to longer README files, but it does seem like most public Cesium projects have shorter ones.

azrogers commented:
Used Doxygen's HTML commands to handle GitHub and Doxygen-specific issues in the Markdown files without creating separate versions of the files. One quirk is that the `[TOC]` command seems to need to come after a paragraph, but not at after the *last* paragraph, to actually work. When it's present, no matter where it is, Doxygen generates the TOC at the top of the file, but that just seems a bit weird to me. Anyways, I added tables of contents to a lot of the documentation, as well as organizing things a bit more in Doxygen and adding the Change Log to the Doxygen output. 

The main remaining thing is - what to do about the full dependency graph? It's too large to view without panning and zooming around, even with my best attempts at styling it to be a bit more clear. No matter what I do I can't seem to get Mermaid to generate a vertical chart instead of a horizontal one, which would be much more convenient for placing on a web page. Doxygen comes with an SVG pan-and-zoom feature, but only for its own generated DOT graphs. Probably the easiest way to go about it is just to link people directly to the SVG itself so they can pan and zoom around it themselves, but that seems a bit rough.

azrogers commented:
Turns out that adding an SVG pan-and-zoom library was only 32kb, so I've gone ahead and done that. Added a Dependencies page listing our third party libraries and put the full dependency graph on that. Also made all the dependency graphs on the namespace pages interactive, so that big ones like the graph for Cesium3DTilesSelection can be viewed more easily. Added a `@mermaid-interactive` alias for these graphs, so regular Mermaid charts that we don't want to be pannable and zoomable aren't affected. I also removed the `layout: elk` line from all the charts but the `all` chart, since it's not really required for any of the other ones and adding the whole `@mermaid-js/layout-elk` library would be a lot of JavaScript for such a small thing. I think that's everything?

kring commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,75 @@
+- Pull request tips
+  - If your pull request fixes an existing issue, include a link to the issue in the description (like this: &quot;Fixes [#1](https://github.com/CesiumGS/cesium-native/issues/1)&quot;). Likewise, if your pull request fixes an issue reported on the Cesium forum, include a link to the thread.
+  - If your pull request needs additional work, include a [task list](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments).
+  - Once you are done making new commits to address feedback, add a comment to the pull request such as `"this is ready"` since GitHub doesn't notify us about commits.
```
I think this was copied from CesiumJS, but GitHub does notify of commits now. I suggest just removing "since GitHub doesn't notify us about commits" entirely.

kring reviewed (commented):
This looks great @azrogers. Just a few more minor things below!

kring commented on README.md:
```diff
@@ -1,125 +1,112 @@
+  <tbody>
+    <tr>
+      <td>
+        <!--! \if DOXYGEN_EXCLUDE -->
+        <img src="doc/img/screenshot-googleplex.jpg" alt="Googleplex" />
+        <!--! \endif -->
+        <!--! \image html screenshot-googleplex.jpg -->
```
You should be able to just use a markdown image rather than these conditionals. I tried it out in Doxygen and it seems to work. I haven't tried on GitHub but I can't think of any reason it wouldn't work there, too.
```suggestion
        ![Googleplex](doc/img/screenshot-googleplex.jpg)
```

azrogers replied:
Unfortunately GitHub is what breaks here. For whatever reason, without some weird spacing tweaks, GitHub will just render the Markdown image tag here as plaintext. And with some weird spacing tweaks, [as Janine noted here](https://github.com/CesiumGS/cesium-native/pull/1024#discussion_r1871540966), it causes noticeable whitespace to show up in the output. But we can't just use the `<img>` tag, because Doxygen won't remap the URL if we do that.

azrogers commented on doc/topics/dependencies.md:
```diff
@@ -0,0 +1,33 @@
+
+The following chart illustrates the connections between the Cesium Native libraries and third-party dependencies:
+
+\image html dependency-graph.svg
```
I can make it larger, which I've done by adding an alias that gives the viewport the specified min-height. For width, though, it's a bit trickier since it's currently limited by the width of the page container. We can get around this by adding a negative margin to the left and right sides, but this will make the pane too large on small screens. Media queries can help with this, but they can't get around the fact that the width of the sidebar can be changed, which means some JavaScript would be needed to adjust the size as it changes. I can add this JavaScript if it seems useful.

azrogers commented on doc/topics/developer-setup.md:
````diff
@@ -0,0 +1,115 @@
+
+```bash
+## Windows compilation using Visual Studio
+cmake -B build -S . -G "Visual Studio 15 2017 Win64"
````
Updated the line to work with Visual Studio 2022 instead, we can do a more thorough pass later.

azrogers commented:
@kring Made some changes based on your review. The main outstanding issue is the sizing on the dependency graph viewport. And if we'd like to try harder to find a way to make the Markdown image syntax work in the screenshots table.

kring commented:
Looks great, thanks @azrogers!
